In [1]:
import pandas as pd
import numpy as np
import pickle

from crossref.restful import Works, Etiquette
from scihub_blastoise import SciHub_HydroPump


In [2]:
with open("q1_jrnl_expladed", "rb") as fp:
    q1_jrnl_expladed = pickle.load(fp)

with open("q1_jrnl_df", "rb") as fp:
    q1_jrnl_df = pickle.load(fp)

__NOTE:__ Before the loop, it is realised that using q1_jrnl_df during the loop will prevent duplicate article extraction by a HUGE margin. therefore, although it will be double-work, it makes much more sense to use it & attach s_a info to articles after the extraction is completed!

TTD:

* Need to convert ISSN colm to list - DONE - in SumCetCet below
* Need to update useful_cols - DONE - only removed "reference" column
* In addition to prob.13, there are 44 journals that exceed their CR_Count when the Current_Count is also considered. Fortunately, this is also a very low minority when the total dataset is considered, but still neeeds to be accounted for! - SKIPPED - will deal with later

In [37]:
# TTD #2:
with open("q1_date_df", "rb") as fp:
    q1_date_df = pickle.load(fp)

print(q1_date_df.columns)

useful_cols = ['DOI', 'references-count', 'publisher', 'published-print', 'is-referenced-by-count', 'title', 'author', 'published-online', 'container-title', 'issued', 'ISSN', 'subject']


Index(['JRNL_ID', 'ISSN', 'issued', 'container-title', 'publisher', 'title',
       'author', 'subject', 'references-count', 'is-referenced-by-count',
       'published', 'published-print', 'published-online', 'direct_url',
       'Dates', 'Keywords', 'Affiliation'],
      dtype='object')


In [ ]:
# TTD #3:
q1_jrnl_df[q1_jrnl_df.Required_Count + q1_jrnl_df.Current_Count > q1_jrnl_df.CR_Count]

## Part 1: CR Run for q1_jrnl_fe

In [30]:
# Test Run:

# SumCetCet CR Loop Run (for q1_jrnl_fe):

from crossref.restful import Works, Etiquette

testo = q1_jrnl_fe.sample(5)
print(testo.Required_Count.sum())


my_etiquette = Etiquette('Analysing Publishing Delay in Academic Journals', 'v2.0', 'https://github.com/Spidey0023/THEsis-Codes', 'oguzkokes@gmail.com')

ogi_works = Works(etiquette=my_etiquette)

q1_dict = testo[["ISSN","Year", "Required_Count"]].to_dict("index")

#Need to convert ISSN column to list!!!


cet_q1_meta_fe = []
useful_cols = ['DOI', 'references-count', 'publisher', 'published-print', 'is-referenced-by-count', 'title', 'author', 'published-online', 'container-title', 'issued', 'ISSN', 'subject']


for key in q1_dict:
    issn = q1_dict[key]["ISSN"]
    year = q1_dict[key]["Year"]
    rc = q1_dict[key]["Required_Count"]

    loop_meta = [w for w in ogi_works.filter(issn=list(issn), from_pub_date=year, until_pub_date=year, type="journal-article").sample(int(rc)).select(useful_cols)] 
    loop_meta = [dict(item, JRNL_ID=key) for item in loop_meta]
        
    cet_q1_meta_fe.extend(loop_meta)


# open("cet_q1_meta_fe","wb") as fp:
#    pickle.dump(cet_q1_meta_fe, fp)

Test run is succesful! Will move on to the SumCetCet run!

In [ ]:
# SumCetCet CR Loop Run (for q1_jrnl_fe):

from crossref.restful import Works, Etiquette
import pandas as pd
import pickle

with open("q1_jrnl_df", "rb") as fp:
    q1_jrnl_df = pickle.load(fp)

# Filtering only necessary jrnls:
q1_jrnl_fe = q1_jrnl_df[q1_jrnl_df.Required_Count >0].copy()
q1_jrnl_fe["ISSN"] = q1_jrnl_fe.ISSN.map(list)

#SumCetCet division:
q1_cr_sum = q1_jrnl_fe[:7000]
q1_cr_cet = q1_jrnl_fe[70000:170000]
q1_cr_ogi = q1_jrnl_fe[170000:]

mail = ""
# 'sduzenli@yandex.com.tr'
# 'cetiner.mserdar@gmail.com'
# 'oguzkokes@gmail.com'

my_etiquette = Etiquette('Analysing Publishing Delay in Academic Journals', 'v2.0', 'https://github.com/Spidey0023/THEsis-Codes', mail)


ogi_works = Works(etiquette=my_etiquette)

q1_dict = q1_cr_cet[["ISSN","Year", "Required_Count"]].to_dict("index")


cet_q1_meta_fe = []
useful_cols = ['DOI', 'references-count', 'publisher', 'published-print', 'is-referenced-by-count', 'title', 'author', 'published-online', 'container-title', 'issued', 'ISSN', 'subject']


for key in q1_dict:
    issn = q1_dict[key]["ISSN"]
    year = q1_dict[key]["Year"]
    rc = q1_dict[key]["Required_Count"]

    loop_meta = [w for w in ogi_works.filter(issn=issn, from_pub_date=year, until_pub_date=year, type="journal-article").sample(rc).select(useful_cols)] 
    loop_meta = [dict(item, JRNL_ID=key) for item in loop_meta]
        
    cet_q1_meta_fe.extend(loop_meta)


with open("cet_q1_meta_fe","wb") as fp:
    pickle.dump(cet_q1_meta_fe, fp)

In [1]:
# Ogi CR Loop Run (for q1_jrnl_fe):

from crossref.restful import Works, Etiquette
import pandas as pd
import pickle

with open("q1_jrnl_df", "rb") as fp:
    q1_jrnl_df = pickle.load(fp)

# Filtering only necessary jrnls:
q1_jrnl_fe = q1_jrnl_df[q1_jrnl_df.Required_Count >0].copy()
q1_jrnl_fe["ISSN"] = q1_jrnl_fe.ISSN.map(list)

#SumCetCet division:
q1_cr_ogi = q1_jrnl_fe[17000:]


mail = 'oguzkokes@gmail.com'
my_etiquette = Etiquette('Analysing Publishing Delay in Academic Journals', 'v2.0', 'https://github.com/Spidey0023/THEsis-Codes', mail)
ogi_works = Works(etiquette=my_etiquette)

q1_dict = q1_cr_ogi[["ISSN","Year", "Required_Count"]].to_dict("index")


ogi_q1_meta_fe = []

useful_cols = ['DOI', 'references-count', 'publisher', 'published-print', 'is-referenced-by-count', 'title', 'author', 'published-online', 'container-title', 'issued', 'ISSN', 'subject']

for key in q1_dict:
    issn = q1_dict[key]["ISSN"]
    year = q1_dict[key]["Year"]
    rc = q1_dict[key]["Required_Count"]

    loop_meta = [w for w in ogi_works.filter(issn=issn, from_pub_date=year, until_pub_date=year, type="journal-article").sample(rc).select(useful_cols)] 
    loop_meta = [dict(item, JRNL_ID=key) for item in loop_meta]
        
    ogi_q1_meta_fe.extend(loop_meta)


with open("ogi_q1_meta_fe","wb") as fp:
    pickle.dump(ogi_q1_meta_fe, fp)

JSONDecodeError: Expecting value: line 1 column 1 (char 0)

An error occured after JRNL 1_49971, will continue loop from this point on!

In [9]:
# DUMPing ogi_q1_meta_fe just in case:

with open("ogi_q1_meta_fe","wb") as fp:
    pickle.dump(ogi_q1_meta_fe, fp)

In [11]:
# Cont. after error:

q1_dict = q1_cr_ogi.loc["1_49971":,["ISSN","Year", "Required_Count"]].to_dict("index")

useful_cols = ['DOI', 'references-count', 'publisher', 'published-print', 'is-referenced-by-count', 'title', 'author', 'published-online', 'container-title', 'issued', 'ISSN', 'subject']

for key in q1_dict:
    issn = q1_dict[key]["ISSN"]
    year = q1_dict[key]["Year"]
    rc = q1_dict[key]["Required_Count"]

    loop_meta = [w for w in ogi_works.filter(issn=issn, from_pub_date=year, until_pub_date=year, type="journal-article").sample(rc).select(useful_cols)] 
    loop_meta = [dict(item, JRNL_ID=key) for item in loop_meta]
        
    ogi_q1_meta_fe.extend(loop_meta)


with open("ogi_q1_meta_fe","wb") as fp:
    pickle.dump(ogi_q1_meta_fe, fp)

NOTE: The "problematic" journal is duplicated, so need to drop half the extracred artlcs after the run is completed!

In [21]:
q1_jrnl_fe[17000:].Required_Count.sum()

42992

## Part 2: SH Run for q1_meta_fe 
__(after SumCetCet CR for q1_jrnl_fe)__

SumCetCet run is complete will now load all fe dfs & analyze results:

In [23]:
with open("sum_q1_meta_fe", "rb") as fp:
    sum_q1_meta_fe = pickle.load(fp)

with open("cet_q1_meta_fe", "rb") as fp:
    cet_q1_meta_fe = pickle.load(fp)

# with open("cet_q1_meta_fe", "rb") as fp:
#     cet_q1_meta_fe = pickle.load(fp)

In [31]:
q1_cr_sum = q1_jrnl_fe[:7000]
q1_cr_cet = q1_jrnl_fe[7000:17000]
# q1_cr_ogi = q1_jrnl_fe[170000:]



In [29]:
# Sum Analysis:

print(f"Required Count: {q1_cr_sum.Required_Count.sum()}")
print(f"Extracted Count: {len(sum_q1_meta_fe)}")

Required Count: 41367
Extracted Count: 41367


In [33]:
# Cet Analysis:

print(f"Required Count: {q1_cr_cet.Required_Count.sum()}")
print(f"Extracted Count: {len(cet_q1_meta_fe)}")
print(f"Difference: {53907-53836}")

Required Count: 53907
Extracted Count: 53836
Difference: 71


In [35]:
# Ogi Analysis:

print(f"Required Count: {q1_cr_ogi.Required_Count.sum()}")
print(f"Extracted Count: {len(ogi_q1_meta_fe)}")
print(f"Difference: {42992-42981}")

Required Count: 42992
Extracted Count: 42981
Difference: 11


In [40]:
q1_meta_fe = pd.concat([pd.DataFrame(sum_q1_meta_fe),pd.DataFrame(cet_q1_meta_fe),pd.DataFrame(ogi_q1_meta_fe)], ignore_index=True)

In [44]:
with open("q1_date_df", "rb") as fp:
    q1_date_df = pickle.load(fp)

Here, it is realized that the S_A column  for q1_date_df was created in (7a) but it was not saved. 

This isnot a huge loss, as we can assign the S_A after the fe SH_run & the entire dataset is collected. This can also allow a better assignment regard Currecnt_Count & Required_Count & CR_Count colunms.

---

The next step is SH_run for fe dataset!

In [53]:
with open("q1_meta_fe","wb") as fp:
    pickle.dump(q1_meta_fe, fp)

In [54]:
q1_meta_fe.shape

(138184, 13)

In [ ]:
# q1_meta_fe SH_Blastoise Ogi Run:

# from scihub_blastoise import SciHub_HydroPump
# import pickle

sh_h = SciHub_HydroPump()

# with open("q1_meta_fe", "rb") as fp:
#     q1_meta_fe = pickle.load(fp)

#q1_meta_fe_sum = q1_meta_fe[:45000]
#q1_meta_fe_cet = q1_meta_fe[45000:95000]
q1_meta_fe_ogi = q1_meta_fe[95000:]


doi_list = q1_meta_fe_ogi.DOI.tolist()

q1_date_fe_ogi = list() 


for doi in doi_list:
    pdf_url, date, kw = sh_h.get_dates_kws(doi)

    q1_date_fe_ogi.append(dict(DOI = doi, direct_url = pdf_url, Dates = date, Keywords = kw))


with open("q1_date_fe_ogi","wb") as fp:
    pickle.dump(q1_date_fe_ogi, fp)


In [62]:
# Test Run:

doi_list = q1_meta_fe.sample(5).DOI.tolist()

q1_date_fe_ogi = list() 


for doi in doi_list:
    pdf_url, date, kw = sh_h.get_dates_kws(doi)

    q1_date_fe_ogi.append(dict(DOI = doi, direct_url = pdf_url, Dates = date, Keywords = kw))


In [66]:
pd.DataFrame(q1_date_fe_ogi)

DOI  \
0     10.1088/0268-1242/30/8/084001   
1         10.1017/s0954579419000130   
2          10.1177/1074248413504994   
3         10.1016/j.whi.2012.10.001   
4  10.1111/j.1755-263x.2010.00162.x   

                                          direct_url  \
0  https://sci.bban.top/pdf/10.1088/0268-1242%252...   
1  https://sci.bban.top/pdf/10.1017/s095457941900...   
2  https://sci.bban.top/pdf/10.1177/1074248413504...   
3  https://sci.bban.top/pdf/10.1016/j.whi.2012.10...   
4  https://sci.bban.top/pdf/10.1111/j.1755-263x.2...   

                                Dates  \
0                 [31 May 2015, None]   
1                       no_date_found   
2     [May 28, 2013, August 15, 2013]   
3       [8 June 2012, 9 October 2012]   
4  [7 October 2010, 16 December 2010]   

                                            Keywords  
0  [(igbts, 0.3377), (mosfets, 0.3217), (igbt, 0....  
1  [(psychoneuroendocrinology, 0.4846), (stressor...  
2  [(atherosclerosis, 0.5629), (atherothrombosis,...  
3  [(contraception, 0.6124), (contraceptive, 0.56...  
4  [(wildlife, 0.6136), (tiger, 0.5459), (tigers,...

Test run is completed! SumCetCet run can be initiated!

As the whole loop might take a while, will start the loop in chunks below:

__NOTE: DON'T FORGET TO KEEP q1_date_fe_ogi & DO NOT RE-CREATE!__

In [73]:
# q1_meta_fe SH_Blastoise Ogi Run: - PART 1

# from scihub_blastoise import SciHub_HydroPump
# import pickle

sh_h = SciHub_HydroPump()

# with open("q1_meta_fe", "rb") as fp:
#     q1_meta_fe = pickle.load(fp)

#q1_meta_fe_sum = q1_meta_fe[:45000]
#q1_meta_fe_cet = q1_meta_fe[45000:95000]
q1_meta_fe_ogi = q1_meta_fe[95000:105000]


doi_list = q1_meta_fe_ogi.DOI.tolist()

q1_date_fe_ogi = list() 


for doi in doi_list:
    pdf_url, date, kw = sh_h.get_dates_kws(doi)

    q1_date_fe_ogi.append(dict(DOI = doi, direct_url = pdf_url, Dates = date, Keywords = kw))


with open("q1_date_fe_ogi_part1_7c","wb") as fp:
    pickle.dump(q1_date_fe_ogi, fp)


In [84]:
# q1_meta_fe SH_Blastoise Ogi Run: - PART 2

# from scihub_blastoise import SciHub_HydroPump
# import pickle

sh_h = SciHub_HydroPump()

# with open("q1_meta_fe", "rb") as fp:
#     q1_meta_fe = pickle.load(fp)

#q1_meta_fe_sum = q1_meta_fe[:45000]
#q1_meta_fe_cet = q1_meta_fe[45000:95000]
q1_meta_fe_ogi = q1_meta_fe[105000:120000]


doi_list = q1_meta_fe_ogi.DOI.tolist()

#q1_date_fe_ogi = list() 


for doi in doi_list:
    pdf_url, date, kw = sh_h.get_dates_kws(doi)

    q1_date_fe_ogi.append(dict(DOI = doi, direct_url = pdf_url, Dates = date, Keywords = kw))


with open("q1_date_fe_ogi_part2_7c","wb") as fp:
    pickle.dump(q1_date_fe_ogi, fp)


In [86]:
# q1_meta_fe SH_Blastoise Ogi Run: - PART 3

# from scihub_blastoise import SciHub_HydroPump
# import pickle

sh_h = SciHub_HydroPump()

# with open("q1_meta_fe", "rb") as fp:
#     q1_meta_fe = pickle.load(fp)

#q1_meta_fe_sum = q1_meta_fe[:45000]
#q1_meta_fe_cet = q1_meta_fe[45000:95000]
q1_meta_fe_ogi = q1_meta_fe[120000:]


doi_list = q1_meta_fe_ogi.DOI.tolist()

#q1_date_fe_ogi = list() 


for doi in doi_list:
    pdf_url, date, kw = sh_h.get_dates_kws(doi)

    q1_date_fe_ogi.append(dict(DOI = doi, direct_url = pdf_url, Dates = date, Keywords = kw))


with open("q1_date_fe_ogi","wb") as fp:
    pickle.dump(q1_date_fe_ogi, fp)


## Part 3: Run Results Analysis:

SumCetCet for Q1_fe is completed! The nb will be reset and results will be analyzed.

Depending on the extracted information, will re-calculate Required_Count & update nb info accodringly!

In [2]:
with open("q1_date_fe_sum", "rb") as fp:
    q1_date_fe_sum = pickle.load(fp)

with open("q1_date_fe_cet", "rb") as fp:
    q1_date_fe_cet = pickle.load(fp)

with open("q1_date_fe_ogi", "rb") as fp:
    q1_date_fe_ogi = pickle.load(fp)

In [3]:
q1_date_fe_sum = pd.DataFrame(q1_date_fe_sum)
q1_date_fe_cet = pd.DataFrame(q1_date_fe_cet)
q1_date_fe_ogi = pd.DataFrame(q1_date_fe_ogi)

q1_date_fe = pd.concat([q1_date_fe_sum,q1_date_fe_cet,q1_date_fe_ogi], ignore_index=True)

In [17]:
q1_date_fe.Dates.map(type).value_counts()

<class 'list'>    108345
<class 'str'>      29839
Name: Dates, dtype: int64

In [18]:
q1_date_fe[q1_date_fe.Dates.map(type)==str].Dates.value_counts()

no_date_found            26748
article_not_in_Scihub     2951
cant_read_pdf              137
direct_url_error             2
pdf_bytes_error              1
Name: Dates, dtype: int64

It is seen that the majority of the errors are caused by "NO DATE". However, to clearly understand this issue, we need JRNL_ID column as well.

In [19]:
with open("q1_meta_fe", "rb") as fp:
    q1_meta_fe = pickle.load(fp)

q1_date_fe = q1_date_fe.merge(q1_meta_fe[["DOI","JRNL_ID"]], how="left", left_on="DOI", right_on="DOI")

In [33]:
q1_anlys = q1_date_fe.groupby("JRNL_ID").Dates.apply(list)
q1_anlys = pd.DataFrame(q1_anlys)

q1_anlys["list_total"] = q1_anlys.Dates.map(len)
q1_anlys["list_succ"] = q1_date_fe.groupby("JRNL_ID").Dates.apply(list).map(lambda x: len([i for i in x if type(i)==list]))
q1_anlys["list_diff"] = q1_anlys["list_total"] - q1_anlys["list_succ"]

In [79]:
# PARTIAL OK:

len(q1_anlys[q1_anlys.list_diff > 0].sort_values("list_diff",ascending=False))

11054

In [78]:
#FULL OK:
len(q1_anlys[q1_anlys.list_diff == 0])

14012

In [80]:
# ZERO OK:
len(len(q1_anlys[q1_anlys.list_succ == 0].sort_values("list_total", ascending=False)))

1976

There are 3 types of jrnls:

- _FULL OK_ : These journals have completely satistfied the Required_count conditions. These jrnls might be used to instead of failed retr. 

- _PARTIAL OK_ : These journals have DATE INFO only for SOME artcls. These jrnls should be subjected to a custom loop to get maximum artcls

- _ZERO OK_ : These jrnls did not have any DATE INFO from extracted artcls although they were marked as SH_Results == True

To get a better understanding of the situation & calc & create new loops, we need import q1_jrnl_df 

In [48]:
with open("q1_jrnl_df", "rb") as fp:
    q1_jrnl_df = pickle.load(fp)

with open("q1_jrnl_expladed", "rb") as fp:
    q1_jrnl_expladed = pickle.load(fp)

In [51]:
q1_jrnl_df = q1_jrnl_df.merge(q1_anlys[["list_total","list_succ","list_diff"]], how="left", right_index=True, left_index=True)

In [90]:
# NOT in CrossRef:
print(f"Number of not included in CR journals: {len(q1_jrnl_df[(q1_jrnl_df.list_total.isna()) & (q1_jrnl_df.Required_Count > 0)])}")


print(f'Number of artcls from these journals: {q1_jrnl_df[(q1_jrnl_df.list_total.isna()) & (q1_jrnl_df.Required_Count > 0)]["Required_Count"].sum()}')




Number of not included in CR journals: 9
Number of artcls from these journals: 78


In [82]:
# CR_Count changed jrnls:

len(q1_jrnl_df[(q1_jrnl_df.Required_Count > q1_jrnl_df.list_total)])

1

In [77]:
my_etiquette = Etiquette('Analysing Publishing Delay in Academic Journals', 'v2.0', 'https://github.com/Spidey0023/THEsis-Codes', 'oguzkokes@gmail.com')
ogi_works = Works(etiquette=my_etiquette)


ogi_works.filter(issn=["10870571", "1552454X"], from_pub_date=2016, until_pub_date=2016, type="journal-article").count() 

1

__The analysis so far shows that:__

We have 2 areas to focus:

___JRNL related issues: (CrossRef)___

- 9 jrnls were removed from CrossRef entirely!

- 1 jrnl, 1_24843, had CR_Count 78, but now only 1 


___ARTCL related issues: (NO DATE & not in SH)___

- 1976 jrnls were marked as SH_Results = True, but had 0 successful retrial => complete failure - ZERO_OK

- 11054 jrnls PARTIAL_OK

- rest worked completely as expected!


---

___TTD___:

1- create new version of q1_date_df,

- save q1_date_fe
- save q1_meta_fe
- filter & inner join together
- modify & feat. eng 
- merge with q1_date_df to create new version!

2- q1_jrnl_df update:

- 9+1 CR_Count, 
    
- 1976 SH_Results

3- def partial func

4- re-check q1_jrnl_df for CR_Count & SH_Results?

## Part 3: Updates & Corrections to re-create q1_jrnl_df & q1_date_df

#### 1 - q1_date_df updates:

In [92]:
#1 q1_date_fe & q1_date_df merging:

# 1.1 - save q1_date_fe:
q1_date_fe_all = q1_date_fe.copy()

with open("q1_date_fe_all","wb") as fp:
   pickle.dump(q1_date_fe_all, fp)

# Saved in 7c SH run folder!

# Also duplicated q1_meta_fe_all & SAVED in same folder!

In [104]:
# 1.2 - Merging the two:
q1_date_fe = q1_meta_fe.merge(q1_date_fe.drop("JRNL_ID", axis=1), how="inner", left_on="DOI", right_on="DOI")

# 1.3 - Filtering to only successful artcls:
q1_date_fe = q1_date_fe[q1_date_fe.Dates.map(type)==list]
q1_date_fe.set_index("DOI", drop=True, inplace=True)

In [110]:
# 1.4 - Feature Eng. : 
def earlier_date_part(row):
    try:
        pp_year = row["published-print"]["date-parts"][0][0]
    except TypeError:
        pp_year = None

    try:   
        po_year = row["published-online"]["date-parts"][0][0]
    except TypeError:
        po_year = None

    if po_year is None:
        return row["published-print"]

    elif pp_year is None:
        return row["published-online"]

    else:
        if pp_year<po_year:
            return row["published-print"]
        elif po_year<pp_year:
            return row["published-online"]
        else:
            try:
                pp_month = row["published-print"]["date-parts"][0][1]
            except IndexError:
                pp_month = 12
            try:
                po_month = row["published-online"]["date-parts"][0][1]
            except IndexError:
                po_month = 12
            if pp_month<po_month:
                return row["published-print"]
            elif po_month<pp_month:
                return row["published-online"]
            else:
                return row["published-online"]


q1_date_fe.loc[:,"published"] = q1_date_fe.apply(earlier_date_part, axis=1)

q1_date_fe = q1_date_fe[["JRNL_ID", "ISSN", "issued", "container-title", "publisher", "title", "author", "subject", "references-count", "is-referenced-by-count", "published", "published-print", "published-online", "direct_url", "Dates", "Keywords"]]


In [116]:
# 1.5 - Merge:

with open("q1_date_df", "rb") as fp:
    q1_date_df = pickle.load(fp)

It is seen that 1671 articles are duplicated btw q1_date_df & q1_jrnl_fe. When analyzed, the two have different values on "Dates" and "Keywords" columns (as expected). As the newer results are to be more precise & old 1671 will be dropped from q1_date_df

__Before the merge, q1_date_df is SAVED as q1_date_df_old in 7c SH folder!!!__

In [122]:
shared_dois = list(set(q1_date_df.index).intersection(set(q1_date_fe.index)))

In [154]:
# Proof that 2 dfs have different "Dates" for duplicate DOIs:
q1_date_df.loc[shared_dois,"Dates"].equals(q1_date_fe.loc[shared_dois,"Dates"])

False

In [158]:
q1_date_df = pd.concat([q1_date_df.drop(shared_dois), q1_date_fe], verify_integrity=True)

In [161]:
# 1.6 - Saving & DUMPing the updated q1_date_df:

with open("q1_date_df","wb") as fp:
   pickle.dump(q1_date_df, fp)

With this process, q1_date_df is up-to-date & ready for further steps.

However, q1_jrnl_df needs to be analyzed & updated (if necessary) as well...

### 2 - q1_jrnl_df updates:

The initial analysis above showed that there were 3 different problems during q1_date_fe run. These are:

- for 9+1 jrnls, the CR Count is either no longer valid OR all artcls have been removed (CR_Count = 0)
    
    _all these jrnls had >1 Req_count meaning there is no reason for a re-trial & these jnrls can be marked as CR_Count = 0 & 1 accordingly_ - __DONE, marked in Consider col__


- for 1976 jrnls, there are no succesful DATE INFO extracted. Some of these jrnls such as 1_54942 had 37 artcls extracted & non had DATE INFO, so it can / should be dropped!
    
    _However, further analysis below show that most of these jrnls had only <3 artcls retrieved, and therefore should be re-tried!!!_ - __DONE, marked in Consider col__

- for 11054 jrnls, the current status is PARTIAL OK, need to fix!

- for JRNL_ID: 1_49971, there are 10 artcls retr., alth. the Req_Count is 5!

In [165]:
q1_jrnl_df.fillna(0)[q1_jrnl_df.fillna(0).Required_Count > q1_jrnl_df.fillna(0).list_total]

ISSN  Year                              Title  \
ID                                                                       
1_18052           (09096396,)  2013                   Wildlife Biology   
1_24838  (10870571, 1552454X)  2010  Journal of Biomolecular Screening   
1_24839  (10870571, 1552454X)  2012  Journal of Biomolecular Screening   
1_24840  (10870571, 1552454X)  2014  Journal of Biomolecular Screening   
1_24841  (10870571, 1552454X)  2011  Journal of Biomolecular Screening   
1_24842  (10870571, 1552454X)  2015  Journal of Biomolecular Screening   
1_24843  (10870571, 1552454X)  2016  Journal of Biomolecular Screening   
1_24844  (10870571, 1552454X)  2013  Journal of Biomolecular Screening   
1_66638  (24726311, 24726303)  2019                    SLAS Technology   
1_66639  (24726311, 24726303)  2020                    SLAS Technology   

                                                  Sbj_Area  Total_Docs  \
ID                                                                       
1_18052                       {'Environmental Science': 1}          51   
1_24838  {'Pharmacology, Toxicology and Pharmaceutics': 1}         132   
1_24839  {'Pharmacology, Toxicology and Pharmaceutics': 1}         139   
1_24840  {'Chemistry': 1, 'Biochemistry, Genetics and M...         139   
1_24841  {'Pharmacology, Toxicology and Pharmaceutics': 1}         129   
1_24842  {'Chemistry': 1, 'Biochemistry, Genetics and M...         125   
1_24843  {'Chemistry': 1, 'Biochemistry, Genetics and M...         113   
1_24844  {'Chemistry': 1, 'Pharmacology, Toxicology and...         118   
1_66638                          {'Health Professions': 1}          66   
1_66639                          {'Health Professions': 1}          73   

               Country            Region  \
ID                                         
1_18052        Denmark    Western Europe   
1_24838  United States  Northern America   
1_24839  United States  Northern America   
1_24840  United States  Northern America   
1_24841  United States  Northern America   
1_24842  United States  Northern America   
1_24843  United States  Northern America   
1_24844  United States  Northern America   
1_66638  United States  Northern America   
1_66639  United States  Northern America   

                                    Publisher   Coverage    SJR  ...  \
ID                                                               ...   
1_18052  Nordic Council for Wildlife Research  1995-2020  0,822  ...   
1_24838                SAGE Publications Inc.  1996-2017  1,017  ...   
1_24839                SAGE Publications Inc.  1996-2017  0,964  ...   
1_24840                SAGE Publications Inc.  1996-2017  1,155  ...   
1_24841                SAGE Publications Inc.  1996-2017  0,987  ...   
1_24842                SAGE Publications Inc.  1996-2017  1,080  ...   
1_24843                SAGE Publications Inc.  1996-2017  1,347  ...   
1_24844                SAGE Publications Inc.  1996-2017  1,000  ...   
1_66638                SAGE Publications Inc.  2017-2020  0,632  ...   
1_66639                SAGE Publications Inc.  2017-2020  0,714  ...   

         Citable Docs. (3years)  Cites / Doc. (2years)  Ref. / Doc.  CR_Count  \
ID                                                                              
1_18052                     130                   1,32        47,47        56   
1_24838                     318                   2,80        26,29       150   
1_24839                     373                   2,43        30,27       155   
1_24840                     381                   2,41        33,05       148   
1_24841                     341                   2,23        27,72       169   
1_24842                     393                   2,47        39,44       111   
1_24843                     379                   2,54        31,75        78   
1_24844                     394                   2,11        32,87       141   
1_66638                     125            

In [192]:
q1_jrnl_df[q1_jrnl_df.list_succ == 0].sort_values("list_total", ascending=False)

ISSN  Year  \
ID                                    
1_54942  (18060374, 21790302)  2019   
1_64747  (22124403, 22124411)  2019   
1_28313           (13411357,)  2016   
1_27702  (11782013, 11769114)  2020   
1_45401  (15536467, 00029459)  2020   
...                       ...   ...   
1_28782           (13538292,)  2012   
1_28689  (13522310, 18732844)  2020   
1_28602           (13510088,)  2015   
1_28575  (13507486, 14698293)  2020   
1_66903           (87563282,)  2019   

                                                     Title  \
ID                                                           
1_54942              Periodico Tche Quimica (discontinued)   
1_64747  Oral Surgery, Oral Medicine, Oral Pathology an...   
1_28313                               Experimental Animals   
1_27702              International Journal of Nanomedicine   
1_45401       American Journal of Pharmaceutical Education   
...                                                    ...   
1_28782                                   Health and Place   
1_28689                            Atmospheric Environment   
1_28602                           Endocrine-Related Cancer   
1_28575  European Review of History/Revue Europeenne d'...   
1_66903                                               Bone   

                                                  Sbj_Area  Total_Docs  \
ID                                                                       
1_54942  {'Multidisciplinary': 1, 'Health Professions': 1}         265   
1_64747                                   {'Dentistry': 1}         194   
1_28313  {'Agricultural and Biological Sciences': 1, 'V...          53   
1_27702  {'Chemical Engineering': 1, 'Materials Science...         681   
1_45401  {'Social Sciences': 1, 'Pharmacology, Toxicolo...         212   
...                                                    ...         ...   
1_28782              {'Social Sciences': 2, 'Medicine': 1}         176   
1_28689  {'Earth and Planetary Sciences': 1, 'Environme...         609   
1_28602  {'Biochemistry, Genetics and Molecular Biology...         128   
1_28575                         {'Arts and Humanities': 1}          41   
1_66903  {'Medicine': 2, 'Biochemistry, Genetics and Mo...         349   

                Country            Region  \
ID                                          
1_54942          Brazil     Latin America   
1_64747   United States  Northern America   
1_28313           Japan    Asiatic Region   
1_27702     New Zealand    Pacific Region   
1_45401   United States  Northern America   
...                 ...               ...   
1_28782  United Kingdom    Western Europe   
1_28689  United Kingdom    Western Europe   
1_28602  United Kingdom    Western Europe   
1_28575  United Kingdom    Western Europe   
1_66903   United States  Northern America   

                                                Publisher  \
ID                                                          
1_54942                                Tche Quimica Group   
1_64747                                      Elsevier USA   
1_28313  International Press Editing Centre Incorporation   
1_27702                           Dove Medical Press Ltd.   
1_45401      American Association of Colleges of Pharmacy   
...                                                   ...   
1_28782                                     Elsevier Ltd.   
1_28689                                     Elsevier Ltd.   
1_28602                               BioScientifica Ltd.   
1_28575                                         Routledge   
1_66903                                     Elsevier Inc.   

                                      Coverage    SJR  ...  \
ID                                                     ...   
1_54942                              2011-2020  0,689  ...   
1_64747                              2012-2020  0,686  ...   
1_28313                   1974-1980, 1995-2020  0,668  ...   
1_27702                              2006-2020  1,245  ..

In [193]:
q1_jrnl_df[q1_jrnl_df.list_succ == 0].list_total.value_counts().sort_index()

1.0     955
2.0     427
3.0     230
4.0     100
5.0      68
6.0      51
7.0      36
8.0      28
9.0      15
10.0     17
11.0     15
12.0      6
13.0      6
14.0      7
15.0      3
16.0      1
17.0      2
18.0      2
19.0      1
20.0      2
22.0      3
37.0      1
Name: list_total, dtype: int64

NOTE: IT IS IMPORTANT TO NOTE THAT, DURING PARTIAL FIX OR OTHER RUNS, IT IS REEEEAAAAALLLYYY CRUCIAL TO CONSIDER "FAILED" DOIs!!!!!

To ensure unnnecessary duplicate record retrival for already failed artcls!

In [195]:
testo1 = set(q1_meta_fe.DOI.tolist()).difference(set(q1_date_df.index))

testo2 = set(q1_meta_fe.DOI.tolist()).difference(set(q1_date_fe.index))

There are 65 DOI records that are included in testo 1 but not in testo2.

This basically means that for 65 artcls, we were able to retrive dates in earlier loops, but was not able to succ. retr. dates in the latest fe run.

When investigated, it is seen that either the link redirects to a different version of the artcl, or PDF received date is not captured.

Won't lose time over this issue, will use testo1 as q1_date_df is THE df for artcls.

In [232]:
succ_doi_dict = q1_date_df.reset_index().groupby("JRNL_ID").DOI.apply(list).to_dict()

fail_doi_dict = q1_meta_fe.set_index("DOI").loc[~q1_meta_fe.set_index("DOI").index.isin(q1_date_df.index)].reset_index().groupby("JRNL_ID")["DOI"].apply(list)

__Updating Current_Count__

In [253]:
q1_jrnl_df["New_Current_Count"] = q1_date_df.groupby("JRNL_ID")["title"].count()

q1_jrnl_df.New_Current_Count.fillna(0, inplace=True)
q1_jrnl_df["New_Current_Count"] = q1_jrnl_df.New_Current_Count.map(int)

Out of the 96 jrnls that did not have any artlcs, there are 18 of them which still are problematic!

6 of them do not have any DOIs retr. for CR & rest 12, did not have any successful DATE extraction. 

Seen below:

In [277]:
len(q1_jrnl_df[(q1_jrnl_df.New_Current_Count  == 0) & (q1_jrnl_df.SH_Results == True) & (q1_jrnl_df.Sbj_Area.map(lambda x: "Medicine" not in x.keys()))])

18

__TTD:__

- .fillna() for 4 cols. - DONE
- Create New_Req_Count & New_Consider - DONE
- drop unrelated cols - DONE
- using New_Req_Count & CR_Count create run partial_fill()

In [285]:
# Creating New_Req_Count

q1_jrnl_df.loc[:,["Current_Count", "Required_Count", "list_total", "list_succ", "list_diff"]] = q1_jrnl_df.loc[:,["Current_Count", "Required_Count", "list_total", "list_succ", "list_diff"]].fillna(0)

for col in ["Current_Count", "Required_Count", "list_total", "list_succ", "list_diff"]:
    q1_jrnl_df[col] = q1_jrnl_df[col].astype(int)

q1_jrnl_df["New_Required_Count"] = q1_jrnl_df.Required_Count - q1_jrnl_df.list_succ

__New_Consider__ column, to be skipped in the second run!

Conditions:

-  Req_Count \> 10 & list_succ == 0 
- SH_Results == False
- list_total == 0 & Req_Count >0



In [292]:
# Consider col:

q1_jrnl_df["New_Consider"] =  q1_jrnl_df.SH_Results

# 9 not in CR journals
q1_jrnl_df.loc[(q1_jrnl_df.list_total == 0) & (q1_jrnl_df.Required_Count >0), "New_Consider"] = False

# 51 (2 from prev. step) no date for more than 10 artlcs:
q1_jrnl_df.loc[(q1_jrnl_df.Required_Count > 10 ) & (q1_jrnl_df.list_succ == 0),"New_Consider"] = False

# Final simplifications:
q1_jrnl_df.drop(["Current_Count","Required_Count","list_total","list_succ","list_diff"], axis=1, inplace=True)

q1_jrnl_df.rename(columns={"New_Current_Count":"Current_Count", "New_Required_Count":"Required_Count", "New_Consider": "Consider"}, inplace=True)


## CHECKPOINT!

Everything df related should be ready for next steps!

- q1_date_df -> all successful results are saved here

- q1_jrnl_df -> updated according to fe run. Added a Consider col, to track which jrnls. should be considered in the next steps!

In [306]:
with open("q1_date_df","wb") as fp:
   pickle.dump(q1_date_df, fp)

with open("q1_jrnl_df","wb") as fp:
   pickle.dump(q1_jrnl_df, fp)

## Part 4: Partial Fill Run on q1_jrnl_df

In [2]:
with open("q1_jrnl_df", "rb") as fp:
    q1_jrnl_df = pickle.load(fp)

with open("q1_date_df", "rb") as fp:
    q1_date_df = pickle.load(fp)

with open("q1_meta_fe", "rb") as fp:
    q1_meta_fe = pickle.load(fp)



In [3]:
succ_doi_dict = q1_date_df.reset_index().groupby("JRNL_ID").DOI.apply(list).to_dict()
fail_doi_dict = q1_meta_fe.set_index("DOI").loc[~q1_meta_fe.set_index("DOI").index.isin(q1_date_df.index)].reset_index().groupby("JRNL_ID")["DOI"].apply(list)

all_doi_dict = { key:succ_doi_dict.get(key,[])+fail_doi_dict.get(key,[]) for key in set(list(succ_doi_dict.keys())+list(fail_doi_dict.keys())) }

In [67]:
# Test to see if all_doi_dict works:

print(len(fail_doi_dict["1_8030"]))
print(len(succ_doi_dict["1_8030"]))
print(len(all_doi_dict["1_8030"]))

1
10
11


Need a partial_fill_run mask:

In [21]:
partial_mask = (q1_jrnl_df.Consider == True) & (q1_jrnl_df.Current_Count < q1_jrnl_df.CR_Count) & (q1_jrnl_df.Required_Count > 0)

In [103]:
# Ogi Loop - Partial Run (for q1_jnrl_df):

sh_h = SciHub_HydroPump()

mail = 'oguzkokes@gmail.com'
my_etiquette = Etiquette('Analysing Publishing Delay in Academic Journals', 'v2.0', 'https://github.com/Spidey0023/THEsis-Codes', mail)
ogi_works = Works(etiquette=my_etiquette)

partial_df = q1_jrnl_df[partial_mask][["ISSN","Year", "Required_Count","CR_Count"]]
partial_df["ISSN"] = partial_df.ISSN.map(list)
pr_dict = partial_df.to_dict("index")

pr_cr_results = []

pr_sh_succ = list() 
pr_sh_fail = list()

useful_cols = ['DOI', 'references-count', 'publisher', 'published-print', 'is-referenced-by-count', 'title', 'author', 'published-online', 'container-title', 'issued', 'ISSN', 'subject']

print("Starting CR Phase:")


for key in pr_dict:
    issn = pr_dict[key]["ISSN"]
    year = pr_dict[key]["Year"]

    rc = pr_dict[key]["Required_Count"]
    crc = pr_dict[key]["CR_Count"]
    min_rc = min(rc*4, crc)

    loop_meta = [w for w in ogi_works.filter(issn=issn, from_pub_date=year, until_pub_date=year, type="journal-article").sample(int(min_rc)).select(useful_cols)]
    loop_meta = [dict(item, JRNL_ID=key) for item in loop_meta if item["DOI"] not in all_doi_dict[key]]
    pr_cr_results.extend(loop_meta)

    req_check = 0

    for artcl in loop_meta:
        if req_check == rc:
            break
        else:
            doi = artcl["DOI"]
            pdf_url, date, kw = sh_h.get_dates_kws(doi)
            if type(date) == list:
                pr_sh_succ.append(dict(DOI = doi, direct_url = pdf_url, Dates = date, Keywords = kw))
                req_check += 1
            else:
                pr_sh_fail.append(dict(DOI = doi, direct_url = pdf_url, Dates = date, Keywords = kw))

    

Starting CR Phase:


UrlSyntaxError: Integer specified as 140 but must be a positive integer less than or equal to 100.

Got an error as the max. sample size can be 100!!!

Therefore, need to fix that in the code & deal with > 100 records on another loop!

However, we can exactly determine where the error occured:

    JRNL_ID 1_4237 has Req_Count > 100, so we can cont. the loop!

In [149]:
# Saving variables in case:

with open("pr_cr_results","wb") as fp:
   pickle.dump(pr_cr_results, fp)

with open("pr_sh_succ","wb") as fp:
   pickle.dump(pr_sh_succ, fp)

with open("pr_sh_fail","wb") as fp:
   pickle.dump(pr_sh_fail, fp)



In [152]:
# Ogi Loop - Partial Run (for q1_jnrl_df) - PART 2:

sh_h = SciHub_HydroPump()

mail = 'oguzkokes@gmail.com'
my_etiquette = Etiquette('Analysing Publishing Delay in Academic Journals', 'v2.0', 'https://github.com/Spidey0023/THEsis-Codes', mail)
ogi_works = Works(etiquette=my_etiquette)

partial_df = q1_jrnl_df[partial_mask][["ISSN","Year", "Required_Count","CR_Count"]]
partial_df["ISSN"] = partial_df.ISSN.map(list)
partial_df = partial_df.loc["1_4237":]
pr_dict = partial_df.to_dict("index")

# pr_cr_results = []

# pr_sh_succ = list() 
# pr_sh_fail = list()

useful_cols = ['DOI', 'references-count', 'publisher', 'published-print', 'is-referenced-by-count', 'title', 'author', 'published-online', 'container-title', 'issued', 'ISSN', 'subject']

print("Starting CR Phase:")


for key in pr_dict:
    issn = pr_dict[key]["ISSN"]
    year = pr_dict[key]["Year"]

    rc = pr_dict[key]["Required_Count"]
    crc = pr_dict[key]["CR_Count"]
    min_rc = min(rc*4, crc, 100)

    loop_meta = [w for w in ogi_works.filter(issn=issn, from_pub_date=year, until_pub_date=year, type="journal-article").sample(int(min_rc)).select(useful_cols)]
    loop_meta = [dict(item, JRNL_ID=key) for item in loop_meta if item["DOI"] not in all_doi_dict[key]]
    pr_cr_results.extend(loop_meta)

    req_check = 0

    for artcl in loop_meta:
        if req_check == rc:
            break
        else:
            doi = artcl["DOI"]
            pdf_url, date, kw = sh_h.get_dates_kws(doi)
            if type(date) == list:
                pr_sh_succ.append(dict(DOI = doi, direct_url = pdf_url, Dates = date, Keywords = kw))
                req_check += 1
            else:
                pr_sh_fail.append(dict(DOI = doi, direct_url = pdf_url, Dates = date, Keywords = kw))

    

Starting CR Phase:


ReadTimeout: HTTPSConnectionPool(host='api.crossref.org', port=443): Read timed out. (read timeout=100)

Another error occured, causing another re-start. This time HTTPS 443 error. Need to start the loop with JNRL ID 1_12543

 However, it is also seen that only around 25% of the partial fill loop is completed in 2 days. Therefore, will resort to SumCetCet!

In [ ]:
# Saving variables in case:

with open("pr_cr_results","wb") as fp:
   pickle.dump(pr_cr_results, fp)

with open("pr_sh_succ","wb") as fp:
   pickle.dump(pr_sh_succ, fp)

with open("pr_sh_fail","wb") as fp:
   pickle.dump(pr_sh_fail, fp)



Also, need to dump all_doi_dict for SumCet:

In [161]:
with open("all_doi_dict","wb") as fp:
   pickle.dump(all_doi_dict, fp)

In [179]:
# Ogi Loop - Partial Run (for q1_jnrl_df) - PART 3 - SumCetCet's Ogi :
# But this time to keep things consistent, the order will be Ogi - Sum - Cet

sh_h = SciHub_HydroPump()

mail = 'oguzkokes@gmail.com'
my_etiquette = Etiquette('Analysing Publishing Delay in Academic Journals', 'v2.0', 'https://github.com/Spidey0023/THEsis-Codes', mail)
ogi_works = Works(etiquette=my_etiquette)

partial_df = q1_jrnl_df[partial_mask][["ISSN","Year", "Required_Count","CR_Count"]]
partial_df["ISSN"] = partial_df.ISSN.map(list)
partial_df = partial_df.loc["1_12543":"1_24715"]
pr_dict = partial_df.to_dict("index")

# pr_cr_results = []

# pr_sh_succ = list() 
# pr_sh_fail = list()

useful_cols = ['DOI', 'references-count', 'publisher', 'published-print', 'is-referenced-by-count', 'title', 'author', 'published-online', 'container-title', 'issued', 'ISSN', 'subject']

print("Starting CR Phase:")


for key in pr_dict:
    issn = pr_dict[key]["ISSN"]
    year = pr_dict[key]["Year"]

    rc = pr_dict[key]["Required_Count"]
    crc = pr_dict[key]["CR_Count"]
    min_rc = min(rc*4, crc, 100)

    loop_meta = [w for w in ogi_works.filter(issn=issn, from_pub_date=year, until_pub_date=year, type="journal-article").sample(int(min_rc)).select(useful_cols)]
    loop_meta = [dict(item, JRNL_ID=key) for item in loop_meta if item["DOI"] not in all_doi_dict[key]]
    pr_cr_results.extend(loop_meta)

    req_check = 0

    for artcl in loop_meta:
        if req_check == rc:
            break
        else:
            doi = artcl["DOI"]
            pdf_url, date, kw = sh_h.get_dates_kws(doi)
            if type(date) == list:
                pr_sh_succ.append(dict(DOI = doi, direct_url = pdf_url, Dates = date, Keywords = kw))
                req_check += 1
            else:
                pr_sh_fail.append(dict(DOI = doi, direct_url = pdf_url, Dates = date, Keywords = kw))

    

Starting CR Phase:


In [188]:
# Saving variables in case:

with open("pr_cr_results","wb") as fp:
   pickle.dump(pr_cr_results, fp)

with open("pr_sh_succ","wb") as fp:
   pickle.dump(pr_sh_succ, fp)

with open("pr_sh_fail","wb") as fp:
   pickle.dump(pr_sh_fail, fp)



NOTE: To restart the PC, every partial fill related variable is saved under 7c_partial_fill / 7c_SumCet_results!

Can move on with the steps below, once this nb is re-opened.

After the third attempt, the run is finally complete! After SumCetCet runs also finish we can check the current status to re-calculate the necessary artcls & jrnls.

__TTD:__

- Collect & concat all suc_retr's from partial fill
- Follow the steps to update q1_date_df
- Calc. & compare with pivots to understand missing artlcs for each pivot cell
- Re-calc. Req_Count colm. based on CR_Count & Consider cols.
- One final extraction loop to fill the missing values

--

- On nb 7d, collect authors & give AUTH_IDs
- Match authors with q1_date_df
- Get author affiliation info

--

- FIN!

- partial_df.loc[:"1_12543"] - FINISHED ALREADY

- partial_df.loc["1_12543":"1_24715"] - OGİ

- partial_df.loc["1_24716":"1_43375"]  -SUM

- partial_df.loc["1_43376":] - CET


__Testing Area__

In [47]:
mail = 'oguzkokes@gmail.com'

my_etiquette = Etiquette('Analysing Publishing Delay in Academic Journals', 'v2.0', 'https://github.com/Spidey0023/THEsis-Codes', mail)


ogi_works = Works(etiquette=my_etiquette)

In [185]:
pd.DataFrame(pr_cr_results)

publisher                            DOI  \
0                       FapUNIFESP (SciELO)  10.1590/0001-3765201920181156   
1                       FapUNIFESP (SciELO)  10.1590/0001-3765201920181130   
2                       FapUNIFESP (SciELO)  10.1590/0001-3765201820180232   
3                       FapUNIFESP (SciELO)  10.1590/0001-3765201920170705   
4                       FapUNIFESP (SciELO)  10.1590/0001-3765201920181330   
...                                     ...                            ...   
48592  Institute of Mathematical Statistics              10.1214/18-ejp205   
48593  Institute of Mathematical Statistics              10.1214/18-ejp137   
48594  Institute of Mathematical Statistics              10.1214/18-ejp175   
48595  Institute of Mathematical Statistics              10.1214/18-ejp189   
48596  Institute of Mathematical Statistics              10.1214/18-ejp208   

       is-referenced-by-count  \
0                           2   
1                           1   
2                           5   
3                           5   
4                           5   
...                       ...   
48592                       4   
48593                       4   
48594                       8   
48595                       3   
48596                       3   

                                                   title  \
0      [Leachate and vinasse used in a biological pro...   
1      [Direct analysis of vicariance in Neotropical ...   
2      [Ecotoxicology assay for the evaluation of env...   
3      [Repercussions of low fructose-drinking water ...   
4      [Resveratrol and resveratrol-hydroxypropyl-β-c...   
...                                                  ...   
48592    [Traffic distributions of random band matrices]   
48593  [Quantitative estimates for the flux of TASEP ...   
48594  [Coalescent results for diploid exchangeable p...   
48595  [On the Liouville heat kernel for $k$-coarse M...   
48596  [On the speed of once-reinforced biased random...   

                                                  author  \
0      [{'given': 'ANDRE C.', 'family': 'BATTISTI', '...   
1      [{'ORCID': 'http://orcid.org/0000-0003-2662-62...   
2      [{'given': 'JEAMYLLE', 'family': 'NILIN', 'seq...   
3      [{'ORCID': 'http://orcid.org/0000-0002-4402-14...   
4      [{'given': 'JENIFER', 'family': 'KOLLING', 'se...   
...                                                  ...   
48592  [{'given': 'Benson', 'family': 'Au', 'sequence...   
48593  [{'given': 'C.', 'family': 'Bahadoran', 'seque...   
48594  [{'given': 'Matthias', 'family': 'Birkner', 's...   
48595  [{'given': 'Jian', 'family': 'Ding', 'sequence...   
48596  [{'given': 'Andrea', 'family': 'Collevecchio',...   

               published-online                             container-title  \
0      {'date-parts': [[2019]]}  [Anais da Academia Brasileira de Ciências]   
1      {'date-parts': [[2019]]}  [Anais da Academia Brasileira de Ciências]   
2      {'date-parts': [[2019]]}  [Anais da Academia Brasileira de Ciências]   
3      {'date-parts': [[2019]]}  [Anais da Academia Brasileira de Ciências]   
4      {'date-parts': [[2019]]}  [Anais da Academia Brasileira de Ciências]   
...                         ...                                         ...   
48592                       NaN         [Electronic Journal of Probability]   
48593                       NaN         [Electronic Journal of Probability]   
48594                       NaN         [Electronic Journal of Probability]   
48595                       NaN         [Electronic Journal of Probability]   
48596                       NaN         [Electronic Journal of Probability]   

                               issued  references-count  \
0            {'date-parts': [[2019]]}                27   
1            {'date-parts': [[2019]]}               101   
2            {'date-parts': [[2019]]}                50   
3            {'date-parts': [[2019]]}                40   
4            {

In [142]:
pd.DataFrame(pr_cr_results)[pd.DataFrame(pr_cr_results).DOI.str.startswith("10.1902/jop.2014")].JRNL_ID.value_counts()

1_4233    25
1_4236    13
Name: JRNL_ID, dtype: int64

In [144]:
pd.DataFrame(pr_cr_results)[pd.DataFrame(pr_cr_results).JRNL_ID.map(lambda x: x in ["1_4231", "1_4233", "1_4236", "1_4232", "1_4228" ,"1_4230"])].JRNL_ID.value_counts()

1_4231    86
1_4233    81
1_4236    22
1_4232    17
1_4228    14
1_4230    14
Name: JRNL_ID, dtype: int64

In [122]:
q1_jrnl_df.loc["1_4236"]

ISSN                            (00223492, 19433670)
Year                                            2014
Title                      Journal of Periodontology
Sbj_Area                            {'Dentistry': 1}
Total_Docs                                       266
Country                                United States
Region                              Northern America
Publisher                 John Wiley &amp; Sons Inc.
Coverage                        1945-1951, 1965-2020
SJR                                            1,440
H index                                          156
Total Docs. (3years)                             627
Total Refs.                                    10028
Total Cites (3years)                            1797
Citable Docs. (3years)                           603
Cites / Doc. (2years)                           2,80
Ref. / Doc.                                    37,70
CR_Count                                         275
SH_Results                                    

In [186]:
pd.DataFrame(pr_sh_succ)

DOI  \
0      10.1590/0001-3765201820180232   
1      10.1590/0001-3765201920170705   
2      10.1590/0001-3765201920181330   
3      10.1590/0001-3765201920180616   
4      10.1590/0001-3765201920190339   
...                              ...   
12200           10.1214/ejp.v19-2791   
12201              10.1214/18-ejp149   
12202              10.1214/18-ejp139   
12203              10.1214/17-ejp117   
12204              10.1214/18-ejp254   

                                              direct_url  \
0      https://sci.bban.top/pdf/10.1590/0001-37652018...   
1      https://sci.bban.top/pdf/10.1590/0001-37652019...   
2      https://sci.bban.top/pdf/10.1590/0001-37652019...   
3      https://sci.bban.top/pdf/10.1590/0001-37652019...   
4      https://sci.bban.top/pdf/10.1590/0001-37652019...   
...                                                  ...   
12200  https://sci.bban.top/pdf/10.1214/EJP.v19-2791....   
12201  https://sci.bban.top/pdf/10.1214/18-EJP149.pdf...   
12202  https://sci.bban.top/pdf/10.1214/18-EJP139.pdf...   
12203  https://sci.bban.top/pdf/10.1214/17-EJP117.pdf...   
12204  https://sci.bban.top/pdf/10.1214/18-ejp254.pdf...   

                                   Dates  \
0        [March 5, 2018, April 30, 2018]   
1      [September 6, 2017, July 9, 2018]   
2       [December 11, 2018, May 2, 2018]   
3         [June 16, 2018, July 30, 2018]   
4          [March 21, 2019, May 3, 2019]   
...                                  ...   
12200               [None, May 27, 2014]   
12201           [None, February 8, 2018]   
12202            [None, January 9, 2018]   
12203           [None, October 16, 2017]   
12204           [None, December 3, 2018]   

                                                Keywords  
0      [(ecotoxicology, 0.4202), (estuary, 0.4026), (...  
1      [(fructosedrinking, 0.4905), (fructose, 0.4617...  
2      [(kinases, 0.4828), (kinase, 0.4648), (insulin...  
3      [(dietary, 0.4723), (probiotics, 0.4157), (pou...  
4      [(mutations, 0.4803), (mutation, 0.4522), (gen...  
...                                                  ...  
12200  [(decay, 0.2729), (inequality, 0.2564), (inequ...  
12201  [(exponentials, 0.3427), (stochastic, 0.3303),...  
12202  [(stochastics, 0.3688), (gaussian, 0.338), (st...  
12203  [(markovian, 0.3197), (fluctuations, 0.3176), ...  
12204  [(wasserstein, 0.4292), (diffusion, 0.413), (b...  

[12205 rows x 4 columns]

In [187]:
pd.DataFrame(pr_sh_fail)

DOI  \
0      10.1590/0001-3765201920181156   
1      10.1590/0001-3765201920181130   
2      10.1590/0001-3765201920180694   
3      10.1590/0001-3765201920180439   
4      10.1590/0001-3765201920180487   
...                              ...   
14825     10.1109/tmech.2019.2944293   
14826     10.1109/tmech.2011.2162075   
14827             10.1214/16-ejp4639   
14828                10.1214/16-ejp4   
14829              10.1214/18-ejp245   

                                              direct_url  \
0                                  article_not_in_Scihub   
1                                  article_not_in_Scihub   
2                                  article_not_in_Scihub   
3                                  article_not_in_Scihub   
4                                  article_not_in_Scihub   
...                                                  ...   
14825  https://sci.bban.top/pdf/10.1109/tmech.2019.29...   
14826  https://sci.bban.top/pdf/10.1109/tmech.2011.21...   
14827                              article_not_in_Scihub   
14828                              article_not_in_Scihub   
14829                              article_not_in_Scihub   

                       Dates  \
0      article_not_in_Scihub   
1      article_not_in_Scihub   
2      article_not_in_Scihub   
3      article_not_in_Scihub   
4      article_not_in_Scihub   
...                      ...   
14825          no_date_found   
14826          no_date_found   
14827  article_not_in_Scihub   
14828  article_not_in_Scihub   
14829  article_not_in_Scihub   

                                                Keywords  
0                                  article_not_in_Scihub  
1                                  article_not_in_Scihub  
2                                  article_not_in_Scihub  
3                                  article_not_in_Scihub  
4                                  article_not_in_Scihub  
...                                                  ...  
14825  [(ieee, 0.6088), (ieeecollabratecorg, 0.5977),...  
14826  [(biomechatronics, 0.5059), (sensors, 0.4733),...  
14827                              article_not_in_Scihub  
14828                              article_not_in_Scihub  
14829                              article_not_in_Scihub  

[14830 rows x 4 columns]

In [83]:
pr_dict

{'1_20420': {'ISSN': ['09603085', '17443571'],
  'Year': 2019,
  'Required_Count': 4,
  'CR_Count': 174},
 '1_39623': {'ISSN': ['15221504', '10400605'],
  'Year': 2019,
  'Required_Count': 1,
  'CR_Count': 191},
 '1_65029': {'ISSN': ['22147160'],
  'Year': 2019,
  'Required_Count': 4,
  'CR_Count': 44},
 '1_13837': {'ISSN': ['03064379'],
  'Year': 2017,
  'Required_Count': 1,
  'CR_Count': 127},
 '1_39166': {'ISSN': ['15205207', '15206106'],
  'Year': 2012,
  'Required_Count': 1,
  'CR_Count': 1703}}

In [82]:
pd.DataFrame(pr_cr_results)

publisher                   published-print  \
0                       Elsevier BV       {'date-parts': [[2019, 9]]}   
1                       Elsevier BV       {'date-parts': [[2019, 7]]}   
2                       Elsevier BV       {'date-parts': [[2019, 5]]}   
3                       Elsevier BV       {'date-parts': [[2019, 1]]}   
4                       Elsevier BV       {'date-parts': [[2019, 3]]}   
5                       Elsevier BV      {'date-parts': [[2019, 11]]}   
6                       Elsevier BV       {'date-parts': [[2019, 1]]}   
7                       Elsevier BV       {'date-parts': [[2019, 7]]}   
8                       Elsevier BV      {'date-parts': [[2019, 11]]}   
9                       Elsevier BV       {'date-parts': [[2019, 1]]}   
10                      Elsevier BV       {'date-parts': [[2019, 7]]}   
11                      Elsevier BV       {'date-parts': [[2019, 5]]}   
12                      Elsevier BV      {'date-parts': [[2019, 11]]}   
13                      Elsevier BV       {'date-parts': [[2019, 7]]}   
14   American Physiological Society    {'date-parts': [[2019, 7, 1]]}   
15   American Physiological Society    {'date-parts': [[2019, 8, 1]]}   
16   American Physiological Society    {'date-parts': [[2019, 1, 1]]}   
17   American Physiological Society    {'date-parts': [[2019, 6, 1]]}   
18                      Elsevier BV          {'date-parts': [[2019]]}   
19                      Elsevier BV          {'date-parts': [[2019]]}   
20                      Elsevier BV          {'date-parts': [[2019]]}   
21                      Elsevier BV          {'date-parts': [[2019]]}   
22                      Elsevier BV          {'date-parts': [[2019]]}   
23                      Elsevier BV          {'date-parts': [[2019]]}   
24                      Elsevier BV          {'date-parts': [[2019]]}   
25                      Elsevier BV          {'date-parts': [[2019]]}   
26                      Elsevier BV          {'date-parts': [[2019]]}   
27                      Elsevier BV      {'date-parts': [[2017, 10]]}   
28                      Elsevier BV      {'date-parts': [[2017, 10]]}   
29                      Elsevier BV       {'date-parts': [[2017, 7]]}   
30                      Elsevier BV       {'date-parts': [[2017, 8]]}   
31  American Chemical Society (ACS)  {'date-parts': [[2012, 12, 13]]}   
32  American Chemical Society (ACS)   {'date-parts': [[2012, 4, 19]]}   
33  American Chemical Society (ACS)   {'date-parts': [[2013, 2, 14]]}   
34  American Chemical Society (ACS)   {'date-parts': [[2012, 7, 12]]}   

                              DOI  is-referenced-by-count  \
0       10.1016/j.fbp.2019.08.011                      12   
1       10.1016/j.fbp.2019.04.006                       5   
2       10.1016/j.fbp.2019.02.003                      18   
3       10.1016/j.fbp.2018.10.004                       4   
4       10.1016/j.fbp.2018.12.006                      17   
5       10.1016/j.fbp.2019.09.016                      20   
6       10.1016/j.fbp.2018.08.009                       4   
7       10.1016/j.fbp.2019.05.008                      31   
8       10.1016/j.fbp.2019.09.013                      11   
9   10.1016/s0960-3085(18)30940-4                       0   
10      10.1016/j.fbp.2019.04.013                       3   
11      10.1016/j.fbp.2019.02.007                       7   
12      10.1016/j.fbp.2019.08.013                       5   
13      10.1016/j.fbp.2019.04.010                      10   
14     10.1152/ajplung.00073.2019                      17   
15     10.1152/ajplung.00020.2019                       4   
16     10.1152/ajplung.00012.2018                      15   
17     10.1152/ajplung.00391.2018                      48   
18      10.1016/j.orp.2019.100120                       2   
19      10.1016/j.orp.2019.100121                      16   
20      10.1016/j.orp.2019.100110                       4   
21      10.1016/j.orp.2019.100114                      

In [79]:
len(pr_sh_results)

35

In [49]:
useful_cols = ['DOI', 'references-count', 'publisher', 'published-print', 'is-referenced-by-count', 'title', 'author', 'published-online', 'container-title', 'issued', 'ISSN', 'subject']

In [50]:
testo = [w for w in ogi_works.filter(issn="1476072X", from_pub_date=2014, until_pub_date=2014, type="journal-article").sample(10).select(useful_cols)] 


In [56]:
testo = [dict(item, JRNL_ID="OK") for item in testo if item["references-count"]>40]
        


In [68]:
testo[0]["DOI"]

'10.1186/1476-072x-13-32'

In [217]:
import requests
import io
from bs4 import BeautifulSoup
from pdfminer.high_level import extract_text


In [224]:
HEADERS = {'User-Agent': 'Mozilla/5.0 (X11; Linux x86_64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/36.0.1985.143 Safari/537.36'}

pdf_req = requests.get("https://sci.bban.top/pdf/10.1016/j.scib.2019.07.021.pdf",headers=HEADERS)

pdf_io = io.BytesIO(pdf_req.content)

pdf_read = extract_text(pdf_io).replace("\n", "")
